### PR TITLE
PYTHON-2457 Test that clients wait 500ms between failed heartbeat checks

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Implements: https://github.com/mongodb/specifications/pull/879/

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
Patch: https://evergreen.mongodb.com/version/5fbc764b0ae606058eff4c97

**Comment by ShaneHarvey** [comment]
This is ready for review although I still need to create a PYTHON ticket to track the work.

**Review by patrickfreed — APPROVED** [review]

File: test/test_streaming_protocol.py [hunk 1]
```diff
@@ -130,54 +130,28 @@ def changed_event(event):
...
-        self.assertLess(len(hb_started_events), 40)
+            start = time.time()
+            client = single_client(
+                appName='SDAMMinHeartbeatFrequencyTest',
+                serverSelectionTimeoutMS=5000)
+            self.addCleanup(client.close)
+            # Force a connection.
+            client.admin.command('ping')
+            duration = time.time() - start
+            self.assertGreaterEqual(duration, 2)
```
Comment (prashantmital) [inline test/test_streaming_protocol.py:153]: Can you add comments here that explain the math? E.g.
```
# isMaster attempted 5 times with 500 ms between checks => duration > 4 * 500ms 
self.assertGreaterEqual(duration, 2)
```

File: test/test_streaming_protocol.py [hunk 2]
```diff
@@ -130,54 +130,28 @@ def changed_event(event):
...
+            self.assertLessEqual(duration, 3.5)
```
Comment (prashantmital) [inline test/test_streaming_protocol.py:154]: How do you arrive at 3.5s?

**Review by prashantmital — CHANGES_REQUESTED** [review]

File: test/test_streaming_protocol.py (hunk 1 above)
Comment (ShaneHarvey) [reply test/test_streaming_protocol.py:167]: Added a link to the spec and an explanation like the previous version of the test.

File: test/test_streaming_protocol.py (hunk 2 above)
Comment (ShaneHarvey) [reply test/test_streaming_protocol.py:168]: That's what the spec says to use: https://github.com/mongodb/specifications/blob/6c5b2ac/source/server-discovery-and-monitoring/server-discovery-and-monitoring-tests.rst#monitors-sleep-at-least-minheartbeatfreqencyms-between-checks

Comment (patrickfreed) [reply test/test_streaming_protocol.py:168]: I added it as a generous upper bound to account for any timing differences in the various test runner implementatinos. Drivers should really take between 2 and 2.5 seconds (depending on when the timer was started and how fast the monitors start up).

**Review by prashantmital — APPROVED** [review]

**Pull request merged**